tokio/src/sync/notify.rs

```diff
@@ -516,19 +526,44 @@ impl Notify {
             return;
         }
 
-        // At this point, it is guaranteed that the state will not
-        // concurrently change, as holding the lock is required to
-        // transition **out** of `WAITING`.
+        // Move all waiters to a list owned by this function call to ensure atomicity.
+        // This list will still be protected by the `waiters` lock.
+        let waiters_to_notify = UnsafeCell::new(std::mem::take(&mut *waiters));
+        pin!(waiters_to_notify);
+
+        // Inform every waiter that the list it is stored in has been moved by storing
+        // a raw pointer to the list. The list is pinned to the stack of this function
+        // call, so a pointer will remain valid.
+        //
+        // Safety: we hold the `waiters` lock.
+        for mut waiter in unsafe { &*waiters_to_notify.get() } {
+            // Safety: we hold the `waiters` lock.
+            let waiter = unsafe { waiter.as_mut() };
+
+            // Safety: address of `waiters_to_notify` is not null.
+            waiter.notify_waiters_queue =
+                unsafe { Some(NonNull::new_unchecked(waiters_to_notify.get())) };
+        }
```

Comment:
I would prefer to create a raw pointer once and reuse it. This has clearer semantics aliasing-wise.

Reply:
Refactored this part.